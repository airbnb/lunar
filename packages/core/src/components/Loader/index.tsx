import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Display inline instead of absolutely positioned. */
  inline?: boolean;
  /** Invert colors. */
  inverted?: boolean;
  /** Increase the dot size. */
  large?: boolean;
  /** Position statically instead of absolutely. */
  static?: boolean;
};

/** A small 3-dot loading indicator. */
export class Loader extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    inline: false,
    inverted: false,
    large: false,
    static: false,
  };

  render() {
    const { cx, styles, inline, inverted, large, static: isStatic } = this.props;

    return (
      <div
        className={cx(
          styles.loader,
          inline && styles.loader_inline,
          !isStatic && !inline && styles.loader_absolute,
        )}
      >
        {[1, 2, 3].map(no => (
          <span
            key={no}
            className={cx(
              styles.dot,
              styles[`dot_${no}`],
              large && styles.dot_large,
              inverted && styles.dot_inverted,
            )}
          />
        ))}
      </div>
    );
  }
}

export default withStyles(({ color, unit }) => ({
  loader: {
    margin: '0 auto',
    textAlign: 'center',
  },

  loader_absolute: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translateX(-50%) translateY(-50%)',
  },

  loader_inline: {
    display: 'inline-block',
  },

  dot: {
    width: 6,
    height: 6,
    marginRight: unit / 2,
    borderRadius: '100%',
    display: 'inline-block',
    animationName: {
      name: 'fade',
      '0%, 80%, 100%': {
        opacity: 0,
      },
      '30%, 50%': {
        opacity: 1,
      },
    },
    animationDuration: '0.8s',
    animationIterationCount: 'infinite',
    animationTimingFunction: 'linear',
    animationFillMode: 'both',
    verticalAlign: 'middle',
    backgroundColor: color.core.primary[3],
  },

  dot_inverted: {
    backgroundColor: color.accent.bg,
  },

  dot_large: {
    width: 12,
    height: 12,
    marginRight: unit,
  },

  dot_1: {
    animationDelay: '-0.3s',
  },

  dot_2: {
    animationDelay: '-0.15s',
  },
}))(Loader);
