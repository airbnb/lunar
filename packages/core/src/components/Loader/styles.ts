import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, unit }) => ({
  loader: {
    margin: '0 auto',
    textAlign: 'center',
  },

  loader_absolute: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translateX(-50%) translateY(-50%)',
  },

  loader_inline: {
    display: 'inline-block',
  },

  dot: {
    width: 6,
    height: 6,
    marginRight: unit / 2,
    borderRadius: '100%',
    display: 'inline-block',
    animationName: {
      name: 'fade',
      '0%, 80%, 100%': {
        opacity: 0,
      },
      '30%, 50%': {
        opacity: 1,
      },
    },
    animationDuration: '0.8s',
    animationIterationCount: 'infinite',
    animationTimingFunction: 'linear',
    animationFillMode: 'both',
    verticalAlign: 'middle',
    backgroundColor: color.core.primary[3],
  },

  dot_inverted: {
    backgroundColor: color.accent.bg,
  },

  dot_large: {
    width: 12,
    height: 12,
    marginRight: unit,
  },

  dot_1: {
    animationDelay: '-0.3s',
  },

  dot_2: {
    animationDelay: '-0.15s',
  },
});

export default styleSheet;

export { styleSheet };
