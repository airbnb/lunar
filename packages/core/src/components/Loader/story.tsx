import React from 'react';
import Text from '../Text';
import Loader from '.';

export default {
  title: 'Core/Loader',
  parameters: {
    inspectComponents: [Loader],
  },
};

export function aLoaderAbsolutelyCenteredWithinAParent() {
  return (
    <div style={{ height: 100, position: 'relative' }}>
      <Loader />
    </div>
  );
}

aLoaderAbsolutelyCenteredWithinAParent.story = {
  name: 'A loader absolutely centered within a parent.',
};

export function aLoaderInline() {
  return (
    <Text>
      Content before the loader. <Loader inline /> Content after the loader.
    </Text>
  );
}

aLoaderInline.story = {
  name: 'A loader inline.',
};

export function withALargerSizePositionedStatically() {
  return <Loader static large />;
}

withALargerSizePositionedStatically.story = {
  name: 'With a larger size positioned statically.',
};
