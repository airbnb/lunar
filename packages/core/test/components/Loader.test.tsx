import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Loader from '../../src/components/Loader';

describe('<Loader />', () => {
  it('renders default', () => {
    const wrapper = shallowWithStyles(<Loader />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inline', () => {
    const wrapper = shallowWithStyles(<Loader inline />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inverted', () => {
    const wrapper = shallowWithStyles(<Loader inverted />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(<Loader large />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders static', () => {
    const wrapper = shallowWithStyles(<Loader static />);

    expect(wrapper).toMatchSnapshot();
  });
});
