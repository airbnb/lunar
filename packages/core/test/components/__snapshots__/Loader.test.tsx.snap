// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> renders default 1`] = `
<div
  className="loader_1pzb0ut-o_O-loader_absolute_1lfa8e4"
>
  <span
    className="dot_1n663f4-o_O-dot_1_q3fdyh"
    key="1"
  />
  <span
    className="dot_1n663f4-o_O-dot_2_1661oda"
    key="2"
  />
  <span
    className="dot_1n663f4"
    key="3"
  />
</div>
`;

exports[`<Loader /> renders inline 1`] = `
<div
  className="loader_1pzb0ut-o_O-loader_inline_36rlri"
>
  <span
    className="dot_1n663f4-o_O-dot_1_q3fdyh"
    key="1"
  />
  <span
    className="dot_1n663f4-o_O-dot_2_1661oda"
    key="2"
  />
  <span
    className="dot_1n663f4"
    key="3"
  />
</div>
`;

exports[`<Loader /> renders inverted 1`] = `
<div
  className="loader_1pzb0ut-o_O-loader_absolute_1lfa8e4"
>
  <span
    className="dot_1n663f4-o_O-dot_1_q3fdyh-o_O-dot_inverted_9isulr"
    key="1"
  />
  <span
    className="dot_1n663f4-o_O-dot_2_1661oda-o_O-dot_inverted_9isulr"
    key="2"
  />
  <span
    className="dot_1n663f4-o_O-dot_inverted_9isulr"
    key="3"
  />
</div>
`;

exports[`<Loader /> renders large 1`] = `
<div
  className="loader_1pzb0ut-o_O-loader_absolute_1lfa8e4"
>
  <span
    className="dot_1n663f4-o_O-dot_1_q3fdyh-o_O-dot_large_1jk0giu"
    key="1"
  />
  <span
    className="dot_1n663f4-o_O-dot_2_1661oda-o_O-dot_large_1jk0giu"
    key="2"
  />
  <span
    className="dot_1n663f4-o_O-dot_large_1jk0giu"
    key="3"
  />
</div>
`;

exports[`<Loader /> renders static 1`] = `
<div
  className="loader_1pzb0ut"
>
  <span
    className="dot_1n663f4-o_O-dot_1_q3fdyh"
    key="1"
  />
  <span
    className="dot_1n663f4-o_O-dot_2_1661oda"
    key="2"
  />
  <span
    className="dot_1n663f4"
    key="3"
  />
</div>
`;
